File: src/python/lib/Lawyer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


from dataclasses import dataclass
from typing import ClassVar
from .worker import Worker


@dataclass
class Lawyer(Worker):
    """Help on the Lawyer module:
    
    Name

        Lawyer 

    
    Description 

        The Lawyer class is developed to simulate a person that is a Lawyer. 

        This class inherits attributes and methods from the Worker class.


        - property decorator: getters for class. 

        - setter decorator: setters for class.

    Args

        first_name: first name of lawyer. 

        last_name: last name of lawyer.

        gender: gender of lawyer.

        height: height of lawyer. 

        weight: weight of lawyer.

        age: age of lawyer. 

        weekly hours: weekly hours of lawyer.

        salary: salary of lawyer.

        law firm: lawfirm of lawyer.


    Usage

         lawyer = Lawyer("Osvaldo", "Aquino", "M", "5'8", 170, 22, 40, 100000, "Vivoni Law Office, LLC")

    
    Package Contents

        - _first_name
        - _last_name
        - _gender
        - _height
        - _age
        - _weekly_hours
        - _salary
        - _person_count 
        - _worker_count
        - _lawyer_count
        - talk()
        - first_name() || first_name(first_name:str)
        - last_name() || last_name(last_name:str)
        - gender() || gender(gender:str)
        - weight() || weight(weight:int)
        - age() || age(age:int)
        - weekly_hours() || weekly_hours(weekly_hours:int)
        - salary() || salary(salary:int)
        - law_firm() || law_firm(law_firm:str)

    Author

        Osvaldo Aquino
    """

    _law_firm: str # represents the law firm of the worker
    _lawyer_count : ClassVar[int] = 0 # static variable used to keep track of each lawyer that is created.


    def __post_init__(self):
        """ Post init function used to increment the global class variables used in this object."""

        Lawyer._person_count+=1
        Lawyer._worker_count+=1
        Lawyer._lawyer_count+=1
        

    @property
    def law_firm(self) -> str:
        """ Returns

                Law firm of the lawyer.

            Usage

                lawyer.law_firm """

        return self._law_firm


    @law_firm.setter
    def law_firm(self, law_firm: str) -> None:
        """ Sets the law firm attribute of the lawyer given a string.

            Args

                law_firm(str): Represents the new law firm that will be set.

            Usage

                lawyer.law_firm = "Vivoni Law Office, LLC" """

        self._law_firm = law_firm


    def talk(self) -> str:
        """ Prints out useful information about the lawyer object.
        
            Usage
            
                lawyer.talk()"""

        print(f'Hello! My name is {self._first_name} {self._last_name}, I am a lawyer that works {self.weekly_hours} a week at {self._law_firm} and I have a salary of {self._salary}.')

File: src/cpp/lib/Lawyer.cpp
/**
 * @file person.hpp
 * @author Osvaldo Aquino (osvaldo.aquino@upr.edu)
 * @brief TODO: Document - Abstract class for person collection.
 * @version 0.1
 * @date 2022-01-10
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include <lib/headers/lawyer.hpp>



Lawyer::Lawyer(std::string fname, std::string lname, std::string g, std::string h, int w, int age, int hours, int salary, std::string law_firm)
{
    this->_first_name = fname;
    this-> _last_name = lname;
    this->_gender = g;
    this->_height = h;
    this->_weight = w;
    this->_age = age;
    this->_weekly_hours = hours;
    this->_law_firm = law_firm;
    _person_count++;
    _worker_count++;
    _lawyer_count++;
}


std::string Lawyer::law_firm(void)
{
    return this->_law_firm;
}


void Lawyer::law_firm(std::string law_firm)
{
    this->_law_firm = law_firm;
}


void Lawyer::talk(void)
{
    std::cout << "Hello! My name is" << this->_first_name << this->_last_name << ", I am a lawyer that works" << this->_weekly_hours
    << "a week at" << this->_law_firm << " and I have a salary of " << this->_salary << "." << std::endl;;
}